src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,395 @@
+	uint64 tableSize = 0;
+	bool check = DistributedTableSize(shardinfo->tableId, TOTAL_RELATION_SIZE, true,
+									  &tableSize);
+	appendStringInfo(AvgHashQuery, "SELECT avg(h)::int,count(*)"
+								   " FROM (SELECT worker_hash(%s) h FROM %s TABLESAMPLE SYSTEM(least(10, 100*10000000000/%lu))"
+								   " WHERE worker_hash(%s)>=%ld AND worker_hash(%s)<=%ld) s",
+					 shardinfo->distributionColumn, shardinfo->tableName,
+					 tableSize,
+					 shardinfo->distributionColumn, shardinfo->shardMinValue,
+					 shardinfo->distributionColumn, shardinfo->shardMaxValue
+					 );
+	ereport(LOG, errmsg("%s", AvgHashQuery->data));
+	SPI_connect();
+	SPI_exec(AvgHashQuery->data, 0);
+	SPITupleTable *tupletable = SPI_tuptable;
+	HeapTuple tuple = tupletable->vals[0];
+	bool isnull;
+	Datum average = SPI_getbinval(tuple, tupletable->tupdesc, 1, &isnull);
+	int64 isResultNull = 1;
+	if (!isnull)
+	{
+		isResultNull = 0;
+	}
+	SPI_freetuptable(tupletable);
+	SPI_finish();
+
+	if (isResultNull == 0)
+	{
+		return DatumGetInt64(average);
+	}
+	else
+	{
+		return shardinfo->shardMinValue - 1;
+	}
+}
+
+
+/*
+ * This function executes a query and then decides whether a shard is subjected for isolation or average hash 2 way split.
+ * If a tenant is found splitpoints for isolation is returned otherwise average hash value is returned.
+ */
+List *
+FindShardSplitPoints(ShardInfo shardinfo)
+{
+	StringInfo CommonValueQuery = makeStringInfo();
+
+	/*
+	 * The inner query for extracting the tenant value having frequency > TenantFrequency is executed on
+	 * every shard of the table and outer query gives the shardid and tenant values as the
+	 * output. Tenant Frequency is a GUC here.
+	 */
+	appendStringInfo(CommonValueQuery,
+					 "SELECT shardid , unnest(result::%s[]) from run_command_on_shards(%s,$$SELECT array_agg(val)"
+					 " FROM pg_stats s , unnest(most_common_vals::text::%s[],most_common_freqs) as res(val,freq)"
+					 " WHERE tablename = %s AND attname = %s AND schemaname = %s AND freq > %f $$)"
+					 " WHERE result <> '' AND shardid = %ld;",
+					 shardinfo->dataType, quote_literal_cstr(shardinfo->tableName),
+					 shardinfo->dataType,
+					 quote_literal_cstr(shardinfo->shardName),
+					 quote_literal_cstr(shardinfo->distributionColumn),
+					 quote_literal_cstr(get_namespace_name(get_rel_namespace(
+															   shardinfo->tableId))),
+					 TenantFrequency,
+					 shardinfo->shardId);
+
+	ereport(LOG, errmsg("%s", CommonValueQuery->data));
+	List *splitPoints = NULL;
+
+	/* Saving the current memory context*/
+	MemoryContext originalContext = CurrentMemoryContext;
+
+	SPI_connect();
+	SPI_exec(CommonValueQuery->data, 0);
+
+	/*Saving the SPI memory context for switching*/
+	MemoryContext spiContext = CurrentMemoryContext;
+
+	int64 rowCount = SPI_processed;
+	int64 average;
+	int32 hashedValue;
+
+	ereport(LOG, errmsg("%ld", rowCount));
+
+	if (rowCount > 0)
+	{
+		SPITupleTable *tupletable = SPI_tuptable;
```

Comment:
Add a comment here, to exaplain how the split points are deducted.

Reply:
explained